pkg/applicationmanager/controllers/applicationbackup.go

```diff
@@ -265,13 +265,52 @@ func (a *ApplicationBackupController) createBackupLocationPath(backup *stork_api
 func (a *ApplicationBackupController) handle(ctx context.Context, backup *stork_api.ApplicationBackup) error {
 	if backup.DeletionTimestamp != nil {
 		if controllers.ContainsFinalizer(backup, controllers.FinalizerCleanup) {
+			// Run the post exec rules if the backup is in ApplicationBackupStageVolumes stage(After the ApplicationBackupStagePreExecRule Stage) AND execRulesCompleted check is negative
+			if backup.Status.Stage == stork_api.ApplicationBackupStageVolumes {
+				log.ApplicationBackupLog(backup).WithField("Event", "Finalizer Cleanup").Infof("BackpCR is marked for deletion during the %v stage", backup.Status.Stage)
+				if terminationChannels, ok := a.terminationChannels[string(backup.UID)]; ok {
+					for _, channel := range terminationChannels {
+						log.ApplicationBackupLog(backup).WithField("Event", "Finalizer Cleanup").Info("Sending termination commands to kill pre-exec pod")
+						channel <- true
+					}
+				}
+				if isexecRulesCompleted, isExists := a.execRulesCompleted[string(backup.UID)]; backup.Spec.PostExecRule != "" && isExists && !isexecRulesCompleted {
+					log.ApplicationBackupLog(backup).WithField("Event", "Finalizer Cleanup").Info("Starting post-exec rule during the backup cr finalizer cleanup")
+					err := a.runPostExecRule(backup)
+					if err != nil && !k8s_errors.IsNotFound(err) {
+						message := fmt.Sprintf("Error running PostExecRule during the finalizer cleanup: %v", err)
+						log.ApplicationBackupLog(backup).WithField("Event", "Finalizer Cleanup").Errorf(message)
+						a.recorder.Event(backup,
+							v1.EventTypeWarning,
+							string(stork_api.ApplicationBackupStatusFailed),
+							message)
+						return fmt.Errorf("%v", message)
+					}
+					a.execRulesCompleted[string(backup.UID)] = true
+				}
+			}
+
+			// Delete post-exec rule CR after executing the postexec rules only for the manual backup as it is managed and created through the px-backup
+			// And also here the deletion will be skipped if the backupCR is created by the schedule or for the restore from the px-backup.
+			if len(backup.Spec.PostExecRule) != 0 && backup.Annotations[utils.PxbackupAnnotationCreateByKey] == utils.PxbackupAnnotationCreateByValue && backup.Annotations["portworx.io/backup-by"] != "backup-schedule" && backup.Annotations["portworx.io/backup-by"] != "restore" {
```

Comment:
can you move this in cleanupResources() to handle all Cr cleanup.
Also for kubevirt support we have added support do delete the ruleCR created by stork. You may want to check if there is no conflict.

Reply:
You already mentioned px-backup will delete the ruleCr so there will be conflict stork deletes the Cr too and pxBackup deletes too. There should be a way to know that delete is triggered by px-backup. In such case, we should avoid deleting the Cr. In my opinion we should defer deleting ruleCr at all unless its created by stork. 
If the applicationBackupCr created by Px-backup but user has issued deleted outside of px-backup, then we should leave the user to delete the Cr as well.